src/os/bluestore/BlueStore.cc

```diff
@@ -12828,20 +12837,19 @@ int BlueStore::_decompress(bufferlist& source, bufferlist* result)
   bluestore_compression_header_t chdr;
   decode(chdr, i);
   int alg = int(chdr.type);
-  CompressorRef cp = compressor;
-  if (!cp || (int)cp->get_type() != alg) {
-    cp = Compressor::create(cct, alg);
-  }
+  CompressorRef cp;
+  cp = alg < int(compressors.size()) ? compressors[alg] : cp;
```

Comment:
If alg >= COMP_ALG_LAST its a data error worth mentioning.
Otherwise we could just `cp = compressors[alg];`.